File: MPC NMPC Agus/Kodingan_MPC_NMPC/NMPC/C++/kode matlab/simulate_nmpc_kapal.m
function [hist_dim, history_input, time_vector, rmse_data] = simulate_nmpc_kapal()
%#codegen

%% Parameter Kapal
L = 101.07;         
B = 14;             
T = 3.7;            
m = 2423*1e3;       
u_0 = 15.4;         
C_B = 0.65;         
x_G = 5.25;         
A_delta = 5.7224;   
rho = 1024;         
r = 0.156*L;        

%% Koefisien Hidrodinamika
Y_v_dot = -(1 + 0.16*C_B*B/T - 5.1*(B/L)^2)*pi*(T/L)^2;
Y_r_dot = -(0.67*(B/L) - 0.0033*(B/T)^2)*pi*(T/L)^2;
N_v_dot = -(1.1*B/L - 0.041*B/T)*pi*(T/L)^2;
N_r_dot = -((1/12) + 0.017*C_B*B/T - 0.33*B/L)*pi*(T/L)^2;
Y_v = -(1 + 0.4*C_B*B/T)*pi*(T/L)^2;
Y_r = -(-0.5 + 2.2*B/L - 0.08*B/T)*pi*(T/L)^2;
N_v = -(0.5 + 2.4*T/L)*pi*(T/L)^2;
N_r = -(0.25 + 0.039*B/T - 0.56*B/L)*pi*(T/L)^2;

m_nd = 2*m/(rho*L^3);
x_G_nd = x_G/L;
I_z_nd = 1.2392*10^(-4);
u_0_nd = 1;

%% Model Gerak Kapal
M_mat = [m_nd-Y_v_dot , m_nd*x_G_nd-Y_r_dot ;
         m_nd*x_G_nd - N_v_dot , I_z_nd - N_r_dot ];

a11 = ((I_z_nd - N_r_dot)*Y_v - (m_nd*x_G_nd - Y_r_dot)*N_v)/det(M_mat);
a12 = ((I_z_nd - N_r_dot)*(Y_r - m_nd*u_0_nd) - (m_nd*x_G_nd - Y_r_dot)*(N_r - m_nd*x_G_nd*u_0_nd))/det(M_mat);
a21 = ((m_nd - Y_v_dot)*N_v - (m_nd*x_G_nd - N_v_dot)*Y_v)/det(M_mat);
a22 = ((m_nd - Y_v_dot)*(N_r - m_nd*x_G_nd*u_0_nd) - (m_nd*x_G_nd - N_v_dot)*(Y_r - m_nd*u_0_nd))/det(M_mat);

A_sys = [a11 , a12 ;
         a21 , a22 ];
B_sys = [0.01 ;
         1];

%% Setup NMPC
Tp = 30;            
T_sim = 1;          
T_sim_total = 150;  
N = round(Tp / T_sim);   

Q = diag([10, 1, 1]);       
R = 1;                      

r_limit = 0.0932; 
r_limit_nd = r_limit * (L / u_0);
u_limit = deg2rad(35);
u_rate_limit = deg2rad(5);
du_max = u_rate_limit * T_sim;

u_prev = 0; 
[A_du, b_du] = du_constraints(N, u_prev, du_max); 

lb = -u_limit * ones(N,1);
ub =  u_limit * ones(N,1);

s0_dim = [0; 0; 0; 100; 0];         
s0_nd = dimensional_to_nondimensional(s0_dim, L, u_0); 
h_ref = [0; 0; 0]; 

% ========================================================
% PREALOKASI MEMORI UNTUK C++ (PENTING)
% ========================================================
num_steps = round(T_sim_total / T_sim);
history_state_nd = zeros(5, num_steps + 1);
history_input = zeros(1, num_steps + 1);

history_state_nd(:, 1) = s0_nd; 
s_nd = s0_nd;

options = optimoptions('fmincon', 'Algorithm', 'sqp', 'Display', 'none', ...
    'MaxIterations', 200, 'OptimalityTolerance', 1e-6);

for k = 1:num_steps
    t = (k-1) * T_sim;
    
    t_pred = t + (1:N)*T_sim;
    x_ref_seq = (h_ref(1) + t_pred' * u_0) / L;  % Pastikan vektor kolom (N x 1)
    y_ref_seq = h_ref(2) / L * ones(N,1);        
    psi_ref_seq = h_ref(3) * ones(N,1);          
    
    cost_fun = @(U) mpc_cost(U, s_nd, u_prev, x_ref_seq, y_ref_seq, psi_ref_seq, T_sim, L, u_0, A_sys, B_sys, u_0_nd, Q, R);
    nonlcon = @(U) state_constraints(U, s_nd, T_sim, L, u_0, A_sys, B_sys, u_0_nd, r_limit_nd);
    
    U0 = u_prev * ones(N,1);
    
    [U_opt, ~, exitflag] = fmincon(cost_fun, U0, A_du, b_du, [], [], lb, ub, nonlcon, options);

    if exitflag <= 0
        % Menghapus warning() karena I/O string kurang direkomendasikan saat compile C++ murni
        U_opt = U0;
    end
    
    u_applied = U_opt(1); 
    history_input(k) = u_applied; 
    
    dt_nd = T_sim * u_0 / L; 
    s_nd = euler_step(@(s,u) ship_dynamics(s, u, A_sys, B_sys, u_0_nd), s_nd, u_applied, dt_nd);
    history_state_nd(:, k+1) = s_nd; 
    
    u_prev = u_applied;
    [A_du, b_du] = du_constraints(N, u_prev, du_max);
end

time_vector = 0:T_sim:T_sim_total;
history_input(num_steps + 1) = history_input(num_steps); % Mengisi input terakhir

%% Konversi State ke Dimensional
hist_dim = zeros(size(history_state_nd)); 
for i = 1:size(history_state_nd,2) 
    hist_dim(:,i) = nondimensional_to_dimensional(history_state_nd(:,i), L, u_0); 
end

%% Perhitungan RMSE
x_ref_full = h_ref(1) + time_vector * u_0;
y_ref_full = h_ref(2) * ones(size(time_vector));
psi_ref_full = h_ref(3) * ones(size(time_vector));

rmse_x = sqrt(mean((hist_dim(3, :) - x_ref_full).^2));
rmse_y = sqrt(mean((hist_dim(4, :) - y_ref_full).^2));
rmse_psi = sqrt(mean((hist_dim(5, :) - psi_ref_full).^2));

rmse_data = [rmse_x, rmse_y, rmse_psi];

end % Akhir dari fungsi utama

%% --- FUNGSI LOKAL ---
function x_next = euler_step(f, x, u, dt)
    x_next = x + dt * f(x, u);
end

function s_dot = ship_dynamics(s, u, A_sys, B_sys, u0_nd)
    v = s(1); r = s(2); psi = s(5);
    v_r_dot = A_sys * [v; r] + B_sys * u;
    x_dot = u0_nd*cos(psi) - v*sin(psi);
    y_dot = u0_nd*sin(psi) + v*cos(psi);
    s_dot = [v_r_dot(1); v_r_dot(2); x_dot; y_dot; r];
end

function J = mpc_cost(U, s0, u_prev, x_ref_seq, y_ref_seq, psi_ref_seq, T_sim, L, u0, A_sys, B_sys, u0_nd, Q, R)
    N = length(U); 
    s = s0;
    dt_nd = T_sim * u0 / L;
    J = 0; 
    
    for i = 1:N
        u = U(i);
        s = euler_step(@(s,u) ship_dynamics(s, u, A_sys, B_sys, u0_nd), s, u, dt_nd);
        err = [s(3) - x_ref_seq(i);
               s(4) - y_ref_seq(i);
               s(5) - psi_ref_seq(i)];
        J = J + err' * Q * err + R * u^2;
    end
end

function [c, ceq] = state_constraints(U, s0, T_sim, L, u0, A_sys, B_sys, u0_nd, r_limit_nd)
    N = length(U);
    s = s0;
    dt_nd = T_sim * u0 / L;
    c = zeros(2*N, 1); % Prealokasi kendala
    
    idx = 1;
    for i = 1:N
        s = euler_step(@(s,u) ship_dynamics(s, u, A_sys, B_sys, u0_nd), s, U(i), dt_nd);
        r = s(2);    
        c(idx) = r - r_limit_nd;
        c(idx+1) = -r_limit_nd - r;
        idx = idx + 2;
    end
    ceq = [];
end

function [A, b] = du_constraints(N, u_prev, du_max)
    A = zeros(2*N, N);
    b = zeros(2*N, 1);
    
    A(1,1) = 1;   b(1) = u_prev + du_max;
    for i = 2:N
        A(i, i-1) = -1;   A(i, i) = 1;
        b(i) = du_max;
    end
    
    A(N+1,1) = -1;   b(N+1) = du_max - u_prev;
    for i = 2:N
        A(N+i, i-1) = 1;   A(N+i, i) = -1;
        b(N+i) = du_max;
    end
end

function x_nd = dimensional_to_nondimensional(x, L, u0)
    x_nd = [x(1)/u0; x(2)*L/u0; x(3)/L; x(4)/L; x(5)];
end

function x_dim = nondimensional_to_dimensional(x, L, u0)
    x_dim = [x(1)*u0; x(2)*u0/L; x(3)*L; x(4)*L; x(5)];
end

File: MPC NMPC Agus/Kodingan_MPC_NMPC/NMPC/C++/codegen/lib/simulate_nmpc_kapal/simulate_nmpc_kapal.cpp
//
// Academic License - for use in teaching, academic research, and meeting
// course requirements at degree granting institutions only.  Not for
// government, commercial, or other organizational use.
//
// simulate_nmpc_kapal.cpp
//
// Code generation for function 'simulate_nmpc_kapal'
//

// Include files
#include "simulate_nmpc_kapal.h"
#include "anonymous_function.h"
#include "fmincon.h"
#include "mean.h"
#include "rt_nonfinite.h"
#include "simulate_nmpc_kapal_data.h"
#include "simulate_nmpc_kapal_initialize.h"
#include "simulate_nmpc_kapal_internal_types1.h"
#include <cmath>
#include <cstring>
#include <emmintrin.h>

// Function Declarations
static void du_constraints(double u_prev, double A[1800], double b[60]);

// Function Definitions
static void du_constraints(double u_prev, double A[1800], double b[60])
{
  std::memset(&A[0], 0, 1800U * sizeof(double));
  A[0] = 1.0;
  b[0] = u_prev + 0.087266462599716474;
  b[30] = 0.087266462599716474 - u_prev;
  for (int i{0}; i < 29; i++) {
    int A_tmp;
    int b_A_tmp;
    A_tmp = i + 60 * i;
    A[A_tmp + 1] = -1.0;
    b_A_tmp = i + 60 * (i + 1);
    A[b_A_tmp + 1] = 1.0;
    b[i + 1] = 0.087266462599716474;
    A[A_tmp + 31] = 1.0;
    A[b_A_tmp + 31] = -1.0;
    b[i + 31] = 0.087266462599716474;
  }
  A[30] = -1.0;
}

void simulate_nmpc_kapal(double hist_dim[755], double history_input[151],
                         double time_vector[151], double rmse_data[3])
{
  static const double s0_nd[5]{0.0, 0.0, 0.0, 0.98941327792618983, 0.0};
  coder::b_anonymous_function cost_fun;
  double A_du[1800];
  double history_state_nd[755];
  double b_y[151];
  double c_y[151];
  double y[151];
  double b_du[60];
  double dv[2];
  double exitflag;
  double u_prev;
  double x_dot_tmp;
  if (!isInitialized_simulate_nmpc_kapal) {
    simulate_nmpc_kapal_initialize();
  }
  //  Parameter Kapal
  //  Koefisien Hidrodinamika
  //  Model Gerak Kapal
  //  Setup NMPC
  u_prev = 0.0;
  du_constraints(0.0, A_du, b_du);
  //  ========================================================
  //  PREALOKASI MEMORI UNTUK C++ (PENTING)
  //  ========================================================
  std::memset(&history_state_nd[0], 0, 755U * sizeof(double));
  for (int i{0}; i < 5; i++) {
    history_state_nd[i] = s0_nd[i];
    cost_fun.workspace.s_nd[i] = s0_nd[i];
  }
  for (int k{0}; k < 150; k++) {
    coder::anonymous_function nonlcon;
    double U_opt[30];
    double b_u_prev[30];
    double b_cost_fun;
    for (int i{0}; i <= 28; i += 2) {
      __m128d r;
      dv[0] = i;
      dv[1] = static_cast<double>(i) + 1.0;
      r = _mm_loadu_pd(&dv[0]);
      _mm_storeu_pd(
          &cost_fun.workspace.x_ref_seq[i],
          _mm_div_pd(
              _mm_mul_pd(
                  _mm_add_pd(_mm_set1_pd((static_cast<double>(k) + 1.0) - 1.0),
                             _mm_add_pd(_mm_set1_pd(1.0), r)),
                  _mm_set1_pd(15.4)),
              _mm_set1_pd(101.07)));
    }
    //  Pastikan vektor kolom (N x 1)
    for (int i{0}; i < 5; i++) {
      nonlcon.workspace.s_nd[i] = cost_fun.workspace.s_nd[i];
    }
    for (int i{0}; i < 30; i++) {
      b_u_prev[i] = u_prev;
    }
    coder::fmincon(cost_fun, b_u_prev, A_du, b_du, nonlcon, U_opt, exitflag);
    if (exitflag <= 0.0) {
      //  Menghapus warning() karena I/O string kurang direkomendasikan saat
      //  compile C++ murni
      for (int i{0}; i < 30; i++) {
        U_opt[i] = u_prev;
      }
    }
    history_input[k] = U_opt[0];
    //  Akhir dari fungsi utama
    //  --- FUNGSI LOKAL ---
    u_prev = cost_fun.workspace.s_nd[0];
    exitflag = std::sin(cost_fun.workspace.s_nd[4]);
    x_dot_tmp = std::cos(cost_fun.workspace.s_nd[4]);
    b_cost_fun = cost_fun.workspace.s_nd[1];
    cost_fun.workspace.s_nd[0] +=
        0.15236964480063322 *
        ((-0.61373138167832486 * cost_fun.workspace.s_nd[0] +
          -0.1017805438031183 * cost_fun.workspace.s_nd[1]) +
         0.01 * U_opt[0]);
    cost_fun.workspace.s_nd[1] +=
        0.15236964480063322 *
        ((-5.0966239122212551 * u_prev +
          -3.4085423899828422 * cost_fun.workspace.s_nd[1]) +
         U_opt[0]);
    cost_fun.workspace.s_nd[2] +=
        0.15236964480063322 * (x_dot_tmp - u_prev * exitflag);
    cost_fun.workspace.s_nd[3] +=
        0.15236964480063322 * (exitflag + u_prev * x_dot_tmp);
    cost_fun.workspace.s_nd[4] += 0.15236964480063322 * b_cost_fun;
    for (int i{0}; i < 5; i++) {
      history_state_nd[i + 5 * (k + 1)] = cost_fun.workspace.s_nd[i];
    }
    u_prev = U_opt[0];
    du_constraints(U_opt[0], A_du, b_du);
  }
  history_input[150] = history_input[149];
  //  Mengisi input terakhir
  //  Konversi State ke Dimensional
  //  Perhitungan RMSE
  for (int i{0}; i < 151; i++) {
    int hist_dim_tmp;
    time_vector[i] = i;
    hist_dim[5 * i] = history_state_nd[5 * i] * 15.4;
    hist_dim_tmp = 5 * i + 1;
    hist_dim[hist_dim_tmp] = history_state_nd[hist_dim_tmp] * 15.4 / 101.07;
    hist_dim_tmp = 5 * i + 2;
    u_prev = history_state_nd[hist_dim_tmp] * 101.07;
    hist_dim[hist_dim_tmp] = u_prev;
    hist_dim_tmp = 5 * i + 3;
    exitflag = history_state_nd[hist_dim_tmp] * 101.07;
    hist_dim[hist_dim_tmp] = exitflag;
    hist_dim_tmp = 5 * i + 4;
    x_dot_tmp = history_state_nd[hist_dim_tmp];
    hist_dim[hist_dim_tmp] = x_dot_tmp;
    u_prev -= 15.4 * static_cast<double>(i);
    y[i] = u_prev * u_prev;
    b_y[i] = exitflag * exitflag;
    c_y[i] = x_dot_tmp * x_dot_tmp;
  }
  rmse_data[0] = std::sqrt(coder::mean(y));
  rmse_data[1] = std::sqrt(coder::mean(b_y));
  rmse_data[2] = std::sqrt(coder::mean(c_y));
}

double simulate_nmpc_kapal_anonFcn1(const double s_nd[5],
                                    const double x_ref_seq[30],
                                    const double U[30])
{
  static const signed char b_iv[9]{10, 0, 0, 0, 1, 0, 0, 0, 1};
  double s[5];
  double b_err[3];
  double err[3];
  double varargout_1;
  for (int i{0}; i < 5; i++) {
    s[i] = s_nd[i];
  }
  varargout_1 = 0.0;
  for (int i{0}; i < 30; i++) {
    double b_s;
    double b_x_dot_tmp;
    double c_s;
    double d;
    double d_s;
    double e_s;
    double f_s;
    double x_dot_tmp;
    //  Akhir dari fungsi utama
    //  --- FUNGSI LOKAL ---
    b_s = s[0];
    c_s = s[1];
    d = U[i];
    d_s = s[4];
    x_dot_tmp = std::sin(s[4]);
    b_x_dot_tmp = std::cos(s[4]);
    e_s = s[2];
    f_s = s[3];
    s[0] = b_s + 0.15236964480063322 *
                     ((-0.61373138167832486 * b_s + -0.1017805438031183 * c_s) +
                      0.01 * d);
    s[1] =
        c_s + 0.15236964480063322 *
                  ((-5.0966239122212551 * b_s + -3.4085423899828422 * c_s) + d);
    e_s += 0.15236964480063322 * (b_x_dot_tmp - b_s * x_dot_tmp);
    s[2] = e_s;
    b_x_dot_tmp = f_s + 0.15236964480063322 * (x_dot_tmp + b_s * b_x_dot_tmp);
    s[3] = b_x_dot_tmp;
    x_dot_tmp = d_s + 0.15236964480063322 * c_s;
    s[4] = x_dot_tmp;
    err[0] = e_s - x_ref_seq[i];
    err[1] = b_x_dot_tmp;
    err[2] = x_dot_tmp;
    std::memset(&b_err[0], 0, 3U * sizeof(double));
    e_s = 0.0;
    for (int b_i{0}; b_i < 3; b_i++) {
      b_s = ((b_err[b_i] + err[0] * static_cast<double>(b_iv[3 * b_i])) +
             b_x_dot_tmp * static_cast<double>(b_iv[3 * b_i + 1])) +
            x_dot_tmp * static_cast<double>(b_iv[3 * b_i + 2]);
      b_err[b_i] = b_s;
      e_s += b_s * err[b_i];
    }
    varargout_1 = (varargout_1 + e_s) + d * d;
  }
  return varargout_1;
}

void simulate_nmpc_kapal_anonFcn2(const double s_nd[5], const double U[30],
                                  double varargout_1[60])
{
  double s[5];
  for (int i{0}; i < 5; i++) {
    s[i] = s_nd[i];
  }
  std::memset(&varargout_1[0], 0, 60U * sizeof(double));
  //  Prealokasi kendala
  for (int i{0}; i < 30; i++) {
    double b_s;
    double c_s;
    double d;
    int varargout_1_tmp;
    //  Akhir dari fungsi utama
    //  --- FUNGSI LOKAL ---
    b_s = s[0];
    c_s = s[1];
    d = U[i];
    s[0] = b_s + 0.15236964480063322 *
                     ((-0.61373138167832486 * b_s + -0.1017805438031183 * c_s) +
                      0.01 * d);
    b_s =
        c_s + 0.15236964480063322 *
                  ((-5.0966239122212551 * b_s + -3.4085423899828422 * c_s) + d);
    s[1] = b_s;
    varargout_1_tmp = i << 1;
    varargout_1[varargout_1_tmp] = b_s - 0.61167038961038955;
    varargout_1[varargout_1_tmp + 1] = -0.61167038961038955 - b_s;
  }
}

// End of code generation (simulate_nmpc_kapal.cpp)
